app/controllers/course_summaries_controller.rb

```diff
@@ -2,17 +2,50 @@ class CourseSummariesController < ApplicationController
   include CourseSummariesHelper
 
   before_action :authorize_only_for_admin,
-                except: [:populate]
+                except: [:populate, :index]
 
   layout 'assignment_content'
 
   def index
   end
 
   def populate
+    averages = {}
+    medians = {}
+    visible_assessment_totals = []
+    if current_user.admin?
+      visible_assessments = Assessment.all.order(id: :asc).pluck_to_hash(:id, :type, :short_identifier)
+      MarkingScheme.all.each do |m|
+        averages[m.name] = DescriptiveStatistics.mean(m.students_weighted_grades_array(current_user)).round(2)
+      end
+    else
+      visible_assessments = Assessment.where(is_hidden: false).order(id: :asc).pluck_to_hash(:id,
+                                                                                             :type,
+                                                                                             :short_identifier)
+    end
+    visible_assessments.each do |a|
+      if a[:type] == 'GradeEntryForm'
+        gef = GradeEntryForm.find(a[:id])
+        visible_assessment_totals << gef.grade_entry_items.sum(:out_of)
+        averages[gef.short_identifier] = gef.calculate_average&.round(2)
+        if current_user.admin?
+          medians[gef.short_identifier] = gef.calculate_median&.round(2)
+        end
+      else
+        assignment = Assignment.find(a[:id])
+        visible_assessment_totals << assignment.max_mark
+        averages[assignment.short_identifier] = assignment.results_average&.round(2)
+        if current_user.admin? || assignment.display_median_to_students
+          medians[assignment.short_identifier] = assignment.results_median&.round(2)
+        end
+      end
+    end
     render json: {
+      averages: averages,
```

Comment:
Don't store separate hashes for averages/medians/totals. Instead, store one hash whose keys are the assessment `short_identifier` and whose values are another hash `{average: ..., median: ..., total: ...}`. (This is a little more compact, and keeps the data grouped in the logical way.)

Reply:
You can create two of these hashes: one for assessments and one for marking schemes.